front: stdcm: update simulation progress button design

SharglutDev opened the pull request:
close #9677 
fix #9322 

To test popup behavior : 
- Animation is not on sketch yet and its duration will be validated with @thibautsailly on November 25th.
When launching a simulation, the launch button should fade out while the simulation progress popup should fade in from the bottom.
- The popup should be sticky to the bottom if it doesn't fit below the form and be fixed to the top when scrolling far bottom (if you have already a simulation in the results, you can toggle the op list to increase the height and test the behavior)

SharglutDev committed:
front: update stdcm simulation progress button design

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: fix stdcm status banner width

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: keep stdcm results displayed while simulation is in progress

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: fix some consist inputs not being disabled while simulation is in progress

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev commented:
@Maymanaf, tagged you so you can check I don't break e2e tests

emersion commented on front/src/applications/stdcm/components/StdcmLoader.tsx:
```diff
@@ -1,42 +1,57 @@
+  useEffect(() => {
+    // Depending on the scroll change the position of the loader between fixed and sticky
+    const handleScroll = () => {
+      const element = document.querySelector('.stdcm-loader') as HTMLElement;
```
This is a bit fragile: React can re-render the element at any time, in which case our manual DOM changes here would be lost.

Would be nicer to add a state storing whether we're scrolled up or not, and the set a class on the loader element based on that state.

SharglutDev replied:
You're right, done

emersion commented on front/src/applications/stdcm/components/StdcmLoader.tsx:
```diff
@@ -1,42 +1,57 @@
+      const element = document.querySelector('.stdcm-loader') as HTMLElement;
+      const scrollTop = window.scrollY;
+
+      if (scrollTop > window.innerHeight - LOADER_OFFSET) {
```
Do we need to have a constant for `LOADER_OFFSET`? Could we use a ref and find the height of the HTML element dynamically instead?

SharglutDev replied:
The threshold where we want to change is position depends of this offset. Also I made a mistake, it's actually not 32 but 32 * 2 (the offset bottom when it's sticky and the offset top when it's fixed). Do you see a patern I'm missing depending of the component's height ?

SharglutDev replied:
I still kept a few `const` like this one. I'm not sure if having a bunch of ref to pass in props would be a better solution. What do you think ?

emersion replied:
In general I dislike hardcoding style-related things in JS, when it really should just be in the CSS. The reason is that (1) it's fragile (depends on font size, padding, margin, text warping, general layout of the page, etc) and (2) it duplicates the information in two spots (CSS and JS).

theocrsb commented on front/src/styles/scss/applications/stdcm/_loader.scss:
```diff
@@ -1,61 +1,67 @@
+  background-color: var(--white100);
+  width: 466px;
+  box-shadow:
+    0px 6px 21px -5px rgba(24, 68, 239, 0.26),
+    0px 16px 30px -5px rgba(0, 0, 0, 0.16),
+    0px 3px 5px -2px rgba(0, 0, 0, 0.1);
```
```suggestion
    0 6px 21px -5px rgba(24, 68, 239, 0.26),
    0 16px 30px -5px rgba(0, 0, 0, 0.16),
    0 3px 5px -2px rgba(0, 0, 0, 0.1);
```

SharglutDev commented:
There is still an issue which needs to be investigated (probably related to your comment @emersion). When the form is long, the threshold for the position switch isn't good anymore because `scrollY` reaches `innerHeight - LOADER_OFFSET * 2` sooner

Maymanaf commented:
> @Maymanaf, tagged you so you can check I don't break e2e tests

Since the E2E tests for LMR run very quickly, I didn't include a check for the simulation progress to avoid flakiness, as the test might fail to detect it.

emersion commented:
Random side note: really not a fan of these manual size computations. Would really prefer if this were just a simple pure CSS `position: sticky` element.

clarani commented on front/src/applications/stdcm/views/StdcmView.tsx:
```diff
@@ -12,13 +12,14 @@ import { replaceElementAtIndex } from 'utils/array';
 import useStdcmForm from '../hooks/useStdcmForm';
 import type { StdcmSimulation } from '../types';
 
+const LAUNCH_BUTTON_FADEOUT_DURATION = 500;
```
Can you add the unit here in comment (I think it should be ms ?)

SharglutDev replied:
Actually I can remove this change (the `setTimeout` in `useEffect`), it used to be for something else but forgot to remove it. I also reremoved the `isPending` to activate the `fade-out` class of the button.

clarani commented on front/src/styles/scss/applications/stdcm/_loader.scss:
```diff
@@ -1,61 +1,102 @@
-  background-color: rgba(255, 255, 255, 1);
-  width: 546px;
+  z-index: 2;
+  padding: 16px 16px 18px;
```
I think we can do this
```suggestion
  padding: 16px 18px;
```

SharglutDev replied:
We can't. `padding: 16px 16px 18px;` stands for 16 top, 16 inline (left and right) and 18 bottom. This won't be the same behavior with `padding: 16px 18px;`

clarani replied:
My bad !

clarani commented on front/src/applications/stdcm/components/StdcmLoader.tsx:
```diff
@@ -1,42 +1,104 @@
+        innerHeight - launchButtonRef.current.getBoundingClientRect().top >
+        LOADER_HEIGHT + LOADER_BOTTOM_OFFSET;
+
+      if (launchButtonRef.current && isLoaderFitting) {
```
You can handle only 3 cases:
```
  if the loader does not fit between the launch button and the bottom of the screen:
    setStatus('fixed-bottom');
    return;
  if the loader should stick to the top of the screen:
    setStatus('fixed-top');
  else:
    setStatus('absolute');
```

SharglutDev replied:
Done

SharglutDev replied:
I pushed another fixup to move the `LoadingStatus` state inside the `StdcmLoader` component. It seems to work the same and doesn't rerender all the form when we scroll.

clarani commented on front/src/applications/stdcm/types.ts:
```diff
@@ -200,3 +200,8 @@ export type StdcmLinkedPathResult = {
 export type ExtremityPathStepType = 'origin' | 'destination';
+
+export type LoaderStatus = {
+  status: 'fixed' | 'sticky' | 'absolute';
```
Maybe we can use `fixed-top` and `fixed-bottom` instead of `fixed` and `sticky`
```suggestion
  status: 'fixed-top' | 'fixed-bottom' | 'absolute';
```

SharglutDev replied:
Done

clarani commented on front/src/styles/scss/applications/stdcm/_loader.scss:
```diff
@@ -1,61 +1,102 @@
+
+  @keyframes slideUp {
+    0% {
+      // Le translate ne doit se passer que si la popup ne rentre pas dans l'espace restant
+      // La map ne doit jamais bouger quand le formulaire grandit ou retrécit ?
```
```suggestion
```

SharglutDev replied:
Done

clarani reviewed (approved):
LGTM :white_check_mark:

theocrsb reviewed (approved):
LGTM and tested

SharglutDev committed:
front: fix stdcm e2e test on tolerance picker

Currently there is no way to close the tolerance picker modal and it can block the next test action.
Add a click on the map to close it before continuing the tests while waiting for the close button to be implemented.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

Maymanaf reviewed (approved):
LGTM for e2e tests

clarani reviewed (approved):
LGTM :white_check_mark:
